content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
+
+MemPass's overhead can be anywhere from a 100x slowdown to over 1000x in the case of Bubblesort. This can be attributed to the fact that MemPass adds instrumentation not only after `malloc` and `free` calls, but also after every `load` and `store`. The benchmarks in question (especially FloatMM, IntMM, and RealMM) do many matrix computations, which end up blowing up the performance cost of our instrumentation significantly. In hindsight, restricting instrumentation to only `malloc` and `free` would significantly reduce the overhead of MemPass. As explained before, we could find some clever way to invalidate pointers once they are freed, thus causing the program to exit when trying to access an invalid location in memory.
+
+### Benchmark Testing: Correctness
+We evaluated the correctness of MemPass on subsets of the Toyota ITC and SARD-100 benchmark tests. The Toyota benchmark tests consist of a family of memory tests, and two test suites that we used are `Double Free` and `Memory Leak` tests.  The Sard-100 benchmark tests are similar, and we used the `cwe-401-memory-leak`, `cwe-415-double-free` and `cwe-416-use-after-free` suites.
+
+>**Note**: We wanted to use the Invalid Memory Access (Use After Free) test set in Toyota as well, but we kept running into segmentation faults with the tests and struggled to debug them.
+
+In general, the`Double Free` and `cwe-415-double-free` benchmark tests consist of normal double free errors, freeing in constant/variable if statements, freeing in a function, freeing in conditional while loops, and freeing in for loops.
+
+The `Memory Leak`, `cwe-401-memory-leak`, and `cwe-416-use-after-free` benchmark tests consist of a series of tests such as allocating memory without freeing, allocating in conditional statements, freeing based on function return values, allocating memory in mutually recursive functions and various branching scenarios.
+
+A comprehensive description of each test case can be found on the respective pages: [Toyota ITC](https://github.com/regehr/itc-benchmarks) and [SARD-100](https://samate.nist.gov/SRD/testsuite.php#sardsuites).
+
+### Benchmark Results
+In [Detection of Security Vulnerabilities in C Code using Runtime Verification](https://nikolai-kosmatov.eu/publications/vorobyov_ks_tap_2018.pdf), the authors provide benchmark test results for E-ACSL, Google Sanitizer, and RV-Match on both the Toyota-ITC and SARD-100 test suites.
+
+| Defect Type          | E-ACSL | Sanitizer | RV-Match |
+|----------------------|--------|-----------|----------|
+| Dynamic Memory Tests | 94%    | 78%       | 94%      |
+
+Unfortunately, the above statistics are not that useful, as our LLVM pass is only able to target double free, memory leaks and use-after-free errors. Still, within the Dynamic Memory tests, we considered the `Double Free` and `Memory Leak` suites of tests. `Double Free` contains 12 cases, and `Memory Leak` contains 18. Over these test suites we achieved the following results:

```

Comment:
I still don't quite understand why these statistics are "not all that useful," but perhaps I would if I knew what the percentages meant or what "Dynamic Memory Tests" are.